Add errors specific to disbursements

codeOfRobin opened the pull request:
#### 27024671470d48970fd3801c8a1ff6dd6bf1ec6c
<pre>
Add errors specific to disbursements

<a href="https://bugs.webkit.org/show_bug.cgi?id=272754">https://bugs.webkit.org/show_bug.cgi?id=272754</a>
<a href="https://rdar.apple.com/118632803">rdar://118632803</a>
Reviewed by Abrar Rahman Protyasha.

Adds the necessary error types for `PKDisbursementError`s

* Source/WebCore/Modules/applepay/ApplePayError.h:
* Source/WebCore/Modules/applepay/ApplePayError.idl:
* Source/WebCore/Modules/applepay/ApplePayErrorCode.h:
* Source/WebCore/Modules/applepay/ApplePayErrorCode.idl:
* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::convertAndValidate):
(WebCore::ApplePaySession::completeShippingContactSelection):
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::appendShippingContactInvalidError):
(WebCore::ApplePayPaymentHandler::retry):
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.mm:
* Source/WebKit/Platform/cocoa/PaymentAuthorizationPresenter.mm:
(WebKit::toPKPaymentErrorCode):
(WebKit::toPKContactField):
(WebKit::toNSError):
(WebKit::PaymentAuthorizationPresenter::completePaymentSession):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/278440@main">https://commits.webkit.org/278440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e58b3b85f8e2ef4d2b041a93b15f23f284392fae

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50542 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29838 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53801 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1232 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52845 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36088 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/882 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41207 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27488 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43517 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22317 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/24885 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/776 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8920 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/46870 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/837 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55390 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25640 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/748 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48617 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43669 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47667 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11073 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27765 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26633 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -419,7 +419,7 @@ ExceptionOr<std::tuple<ApplePayLineItem, Vector<ApplePayLineItem>>> ApplePayPaym
 {
     if (!message.isNull())
-        errors.append(ApplePayError::create(ApplePayErrorCode::ShippingContactInvalid, WTFMove(contactField), WTFMove(message)));
+        errors.append(ApplePayError::create(ApplePayErrorCode::ShippingContactInvalid, WTFMove(contactField), WTFMove(message), std::nullopt));
```
Do you mind making all the calls to `ApplePayError::create()` consistent? (with respect to `std::nullopt` vs `{ }`)

aprotyas replied:
FWIW, in WebKit we're leaning towards the latter, though that isn't ratified in a style guide or anything.

aprotyas commented on Source/WebCore/Modules/applepay/ApplePayError.h:
```diff
@@ -38,9 +38,14 @@ namespace WebCore {
 public:
-    static Ref<ApplePayError> create(ApplePayErrorCode code, std::optional<ApplePayErrorContactField> contactField, const String& message)
+
+    enum class Domain : uint8_t {
```
I might have asked this in the past, but can we make this a `bool` backed type for tighter packing?

codeOfRobin replied:
We did resolve this one earlier :D

aprotyas reviewed (approved):
LGTM, do you mind rebasing against ToT for a fresh EWS run and then I can merge?

Commit:
Add errors specific to disbursements

https://bugs.webkit.org/show_bug.cgi?id=272754
rdar://118632803
Reviewed by Abrar Rahman Protyasha.

Adds the necessary error types for `PKDisbursementError`s

* Source/WebCore/Modules/applepay/ApplePayError.h:
* Source/WebCore/Modules/applepay/ApplePayError.idl:
* Source/WebCore/Modules/applepay/ApplePayErrorCode.h:
* Source/WebCore/Modules/applepay/ApplePayErrorCode.idl:
* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::convertAndValidate):
(WebCore::ApplePaySession::completeShippingContactSelection):
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::appendShippingContactInvalidError):
(WebCore::ApplePayPaymentHandler::retry):
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.mm:
* Source/WebKit/Platform/cocoa/PaymentAuthorizationPresenter.mm:
(WebKit::toPKPaymentErrorCode):
(WebKit::toPKContactField):
(WebKit::toNSError):
(WebKit::PaymentAuthorizationPresenter::completePaymentSession):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/278440@main
